src/main/java/com/vmware/vipclient/i18n/messages/service/ProductService.java

```diff
@@ -58,23 +56,17 @@ public ProductService(BaseDTO dto) {
     /**
      * Retrieves the list of components of a product. It recursively applies data source fallback mechanism in case of failure.
      *
-     * @param msgSourceQueueIter Iterator of DataSourceEnum sources
      * @return list of components of the product specified in the dto object
      */
-    public List<String> getComponents (Iterator<DataSourceEnum> msgSourceQueueIter) {
-        if (!msgSourceQueueIter.hasNext())
-            return null;
-
-        DataSourceEnum dataSource = msgSourceQueueIter.next();
-        ProductOpt opt = dataSource.createProductOpt(dto);
-        List<String> components = opt.getComponents();
-        // If failed to get components from the data source
-        if (components.isEmpty()) {
-            // Try the next dataSource in the queue
-            if (msgSourceQueueIter.hasNext()) {
-                components = getComponents(msgSourceQueueIter);
-                // If no more data source in queue, log the error. This means that neither online nor offline fetch succeeded.
-            } else {
+    public List<String> getComponents(){
+        List<String> components = null;
+        Iterator<DataSourceEnum> msgSourceQueueIter = VIPCfg.getInstance().getMsgOriginsQueue().iterator();
+        while((components == null || components.isEmpty()) && msgSourceQueueIter.hasNext()){
+            DataSourceEnum dataSource = msgSourceQueueIter.next();
+            ProductOpt opt = dataSource.createProductOpt(dto);
+            components = opt.getComponents();
+            // If failed to get components from the data source, log the error.
+            if (components.isEmpty()) {
```

Comment:
```suggestion
            if (components == null || components.isEmpty()) {
```

Reply:
Actually ‘components’ won't be null as opt.getComponents() method new ArrayList and won't return null.